### ISSUE TITLE
[Bug]: Range Search Limitied to 6400 Results

### ISSUE THREAD

**Issue by hakan458** [milvus-io/milvus#28810]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.3.2
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): kafka
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus==2.3.2
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

When using range search with search param `radius` , Milvus only returns up to 6400 results even though topk/limit can be up to 16,384

### Expected Behavior

I expect to get up to 16384 results, if those results are within the `range_filter` and `radius` 

### Steps To Reproduce

```markdown
I have a collection with around 1 million records

If I perform a search with no range filter and limit=16384 , I get 16384 results

results = c.search(
    data=[ref_embd],
    #expr="candidate_task_id != ' '",
    anns_field='img',
    param={'params': {}, 'metric_type': 'L2'},
    limit=16384,
    offset=0,
    output_fields=['id']
)[0]

ids = [result.id for result in results]
len(ids)
>> 16384
```

If I now add `radius` into `params` , while keeping `limit=16384` , I only get 6400 results. 
```
results = c.search(
    data=[ref_embd],
    #expr="candidate_task_id != ' '",
    anns_field='img',
    param={'params': {'radius': 10}, 'metric_type': 'L2'},
    limit=16384,
    offset=0,
    output_fields=['id']
)[0]

ids = [result.id for result in results]
len(ids)
>> 6400
```

I have verified that this happens for ALL collections I have. 

I have also verified that when I get 16384 records, the largest distance value is < 1.0 , so using radius of 10.0 should still get me back all 16384 results

Another note is search speed is orders of magnitude slower when using range search which I dont understand. 

For example when I am using a smaller limit like 30, this same search without `radius` will run in around 120ms , and when I add `radius` to params it takes 7 seconds!!
```


### Milvus Log

_No response_

### Anything else?

_No response_

**Comment by hakan458** [comment]
Ok somehow this is only happening with some collections, not all. I just created a new one, with the exact same data as one that DOES have this issue, and both the 6400 problem, and the speed issue do not happen on this new collection. Looking at the two I have no idea how this is possible, as they are the exact same. 

```
slow_collection = Collection("De65b1f91_87d2_475e_a6c2_83ade2532baf")
fast_collection = Collection("D47868955_9e67_493e_bc7b_a4e86459d6a0")

slow_collection.num_entities
>> 17760
fast_collection.num_entities
>> 17760

slow_collection

<Collection>:
-------------
<name>: De65b1f91_87d2_475e_a6c2_83ade2532baf
<description>: 
<schema>: {'auto_id': True, 'description': '', 'fields': [{'name': 'id', 'description': '', 'type': <DataType.INT64: 5>, 'is_primary': True, 'auto_id': True}, {'name': 'sync_id', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 50}}, {'name': 'candidate_task_id', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 500}}, {'name': 'storage_id', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 500}, 'is_partition_key': True}, {'name': 'img', 'description': '', 'type': <DataType.FLOAT_VECTOR: 101>, 'params': {'dim': 512}}]}

fast_collection
<name>: D47868955_9e67_493e_bc7b_a4e86459d6a0
<description>: 
<schema>: {'auto_id': True, 'description': '', 'fields': [{'name': 'id', 'description': '', 'type': <DataType.INT64: 5>, 'is_primary': True, 'auto_id': True}, {'name': 'sync_id', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 50}}, {'name': 'candidate_task_id', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 500}}, {'name': 'storage_id', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 500}, 'is_partition_key': True}, {'name': 'img', 'description': '', 'type': <DataType.FLOAT_VECTOR: 101>, 'params': {'dim': 512}}]}

%%time

slow_results = slow_collection.search(
    data=[ref_embd],
    #expr="candidate_task_id != ' '",
    anns_field='img',
    param={'params': {'range_filter': 0, 'radius': 10}, 'metric_type': 'L2'},
    limit=16384,
    offset=0,
    output_fields=['id']
)[0]

CPU times: user 13.8 ms, sys: 0 ns, total: 13.8 ms
Wall time: 7.31 s

%%time

fast_results = fast_collection.search(
    data=[ref_embd],
    #expr="candidate_task_id != ' '",
    anns_field='img',
    param={'params': {'range_filter': 0, 'radius': 10}, 'metric_type': 'L2'},
    limit=16384,
    offset=0,
    output_fields=['id']
)[0]

CPU times: user 9.06 ms, sys: 0 ns, total: 9.06 ms
Wall time: 340 ms

slow_ids = [result.id for result in slow_results]
fast_ids = [result.id for result in fast_results]

len(slow_ids)
>> 6400

len(fast_ids)
>> 6400

slow_distances = [result.distance for result in slow_results]
fast_distances = [result.distance for result in fast_results]
slow_distances[:6400] == fast_distances[:6400]
>> True
```

**Comment by yanliang567** [comment]
@hakan458 which index type are you running? could you please provide the full milvus logs?  please refer this [doc](https://github.com/milvus-io/milvus/tree/master/deployments/export-log) to export the whole Milvus logs for investigation.
Also please attach the etcd backup which would help us understand the "slow issue". Check this: https://github.com/milvus-io/birdwatcher for details about how to backup etcd with birdwatcher.

/assign @hakan458

**Comment by hakan458** [comment]
We are using DiskANN index type for all of them. Very very weird that 2 collections that are practically identical behave differently! I created another one to be sure, and this one is also fine. But many collections from the past few days are in this state of only returning 6400 results when radius is used, and very slow. 

I will try to export logs tool now. I might not have permissions to our cluster to be able to run birdwatcher but will try that as well and report back.

**Comment by hakan458** [comment]
[milvus-log.tar.gz](https://github.com/milvus-io/milvus/files/13495036/milvus-log.tar.gz)
Here are logs gathered from that tool. Let me know if you need anything different

**Comment by hakan458** [comment]
[milvus-log.tar.gz](https://github.com/milvus-io/milvus/files/13495059/milvus-log.tar.gz)
I exported again this time with etcd and kafka as well

**Comment by yanliang567** [comment]
@pakelley @hakan458 it seems that there are 240+ collections in the milvus. could you please share the exact 'slow collection' name and 'fast collection' name, so that you would track the 2 collects to investigate the issue?

**Comment by pakelley** [comment]
@hakan458 can correct me if I'm wrong, but from his code I think the slow collection name is `De65b1f91_87d2_475e_a6c2_83ade2532baf` and the fast collection name is `D47868955_9e67_493e_bc7b_a4e86459d6a0`.

**Comment by hakan458** [comment]
@pakelley is correct on the collection names 
New findings today. The "fast collection" which was fast yesterday... is now just as slow as "slow collection" from yesterday. 

Also both of them are now returning 6400 records when using range search. 

So something is happening over time that is causing a degradation somehow... please let us know if you know how this is possible and if a quick fix is possible. 

Attached new logs. I just did a search on the slow and "fast" collections 
[milvus-log.tar.gz](https://github.com/milvus-io/milvus/files/13503960/milvus-log.tar.gz)

**Comment by yanliang567** [comment]
/assign @congqixia 
could you please also take a look?

**Comment by hakan458** [comment]
Once again to summarize

At first when we make these collections - everything is fine. We request limit=16384 while using range search, and we get 16384. The speed is also fast. 

After an unknown amount of time, maybe a few hours, maybe half a day, it gets around 10x slower to perform the same query AND we start to only get 6400 records, even though none of the search parameters, or data in the collection, has changed.

**Comment by liliu-z** [comment]
> Once again to summarize
> 
> At first when we make these collections - everything is fine. We request limit=16384 while using range search, and we get 16384. The speed is also fast.
> 
> After an unknown amount of time, maybe a few hours, maybe half a day, it gets around 10x slower to perform the same query AND we start to only get 6400 records, even though none of the search parameters, or data in the collection, has changed.

For this fast/slow case, can I ask how many records do you have in Milvus?

**Comment by hakan458** [comment]
> > Once again to summarize
> > At first when we make these collections - everything is fine. We request limit=16384 while using range search, and we get 16384. The speed is also fast.
> > After an unknown amount of time, maybe a few hours, maybe half a day, it gets around 10x slower to perform the same query AND we start to only get 6400 records, even though none of the search parameters, or data in the collection, has changed.
> 
> For this fast/slow case, can I ask how many records do you have in Milvus?

Yes all of the details can be seen above in my code block including num entities, the collection details, etc
In this case both of them have 17760. However we are observing this issue with ALL collections. For example there is one with 500k, one with 1 million, etc. All have the issue of getting 6400 records, and all of them are very slow to search when using radius param, but only after some time.

**Comment by pakelley** [comment]
@liliu-z yep, that's correct. The example dataset in Hakan's code (which has 17,760 records) was "fast" 2 days ago, and yesterday became "slow" and would only return 6400 records.

**Comment by liliu-z** [comment]
@hakan458 @pakelley 
First let me try to explain how Milvus' insertion works:
1. The newly inserted data is maintained in a growing segment for better data freshness. The growing segment is currently an in-memory IVFFlat index.
2. Milvus will seal these data into sealed segments and asynchronously build an index upon them.
3. Milvus will asynchronously compact small segments.

Here are my suspicions:

1. For the 17K case, initially, your search might be applied to the growing in-memory IVFFlat, which is fast. After this segment gets sealed and turns into DiskANN, it becomes significantly slower. You can verify this by checking the status of the segment using Grafana/birdwatcher.
2. For the larger case, there might be two reasons:
  1. The same as mentioned above.
  2. Initially, you have many small segments, and a single search can be applied concurrently to multiple segments. After merging into one segment, a single search can only be executed sequentially (but with higher QPS if you try to improve the client concurrency). Also, small segments have fewer vectors satisfying the radius, so the searches applied to each small segment are lighter.

Regarding the 6400 issue, there is a range cap in DiskANN that only allows 6400 results for each segment. This limit will be removed in the next version. This can also an evidence that after merging, you end up with a single large segment.

**Comment by xiaofan-luan** [comment]
Is is more reasonable to simply use IVF_PQ or HNSW_PQ/SQ for such use cases?

I wonder the benefit for the having DiskANN on such a large result set.

**Comment by pakelley** [comment]
> When it get sealed, the reason why it is slow is because you do sequential search upon it. Actually its QPS will increase at the segment number get smaller.
1. Is there anything we can do for now to improve performance with DiskANN?
2. Should we expect [this PR](https://github.com/zilliztech/knowhere/pull/237) to help? (Seems like maybe, if anything, it might return the correct number of records? But not necessarily improve performance?)
3. @xiaofan-luan mentioned HNSW_PQ/SQ, but it seems these are [no longer available](https://github.com/milvus-io/milvus/pull/18935). Am I understanding that correctly?

**Comment by liliu-z** [comment]
> > When it get sealed, the reason why it is slow is because you do sequential search upon it. Actually its QPS will increase at the segment number get smaller.
> 
> 1. Is there anything we can do for now to improve performance with DiskANN?
> 2. Should we expect [this PR](https://github.com/zilliztech/knowhere/pull/237) to help? (Seems like maybe, if anything, it might return the correct number of records? But not necessarily improve performance?)
> 3. @xiaofan-luan mentioned HNSW_PQ/SQ, but it seems these are [no longer available](https://github.com/milvus-io/milvus/pull/18935). Am I understanding that correctly?

Hi @pakelley 
1. Algorithm need to be refactoring and we are working on it. Change this [default value](https://github.com/zilliztech/knowhere/blob/main/src/index/diskann/diskann_config.h#L140-L144) into 1.0 can help accelerate it, with a little sacrifice of performance.
2. This can only help on the number get returned
3. Yes

**Comment by hakan458** [comment]
> > > When it get sealed, the reason why it is slow is because you do sequential search upon it. Actually its QPS will increase at the segment number get smaller.
> > 
> > 
> > 
> > 1. Is there anything we can do for now to improve performance with DiskANN?
> > 2. Should we expect [this PR](https://github.com/zilliztech/knowhere/pull/237) to help? (Seems like maybe, if anything, it might return the correct number of records? But not necessarily improve performance?)
> > 3. @xiaofan-luan mentioned HNSW_PQ/SQ, but it seems these are [no longer available](https://github.com/milvus-io/milvus/pull/18935). Am I understanding that correctly?
> 
> Hi @pakelley
> 
>     1. Algorithm need to be refactoring and we are working on it. Change this [default value](https://github.com/zilliztech/knowhere/blob/main/src/index/diskann/diskann_config.h#L140-L144) into 1.0 can help accelerate it, with a little sacrifice of performance.
> 
>     2. This can only help on the number get returned
> 
>     3. We removed the old implementation of it and working on a new one. Still need some time

Is this value exposed in the config file? Or is only modifiable in the source code itself?

**Comment by yanliang567** [comment]
@liliu-z so any conclusion for this issue?

**Comment by liliu-z** [comment]
> @liliu-z so any conclusion for this issue?

https://github.com/milvus-io/milvus/issues/28810#issuecomment-1839905589
1 and 2 are solved. 3 is a long term work. OK to close
